pypit/arproc.py

```diff
@@ -22,195 +23,95 @@
-# 	arutils.ds9plot(test_rect)
-# 	for o in range(norders):
-# 		pass
-
-
-    assert(False)
-    # Mask out ordpix pixels where there is target flux
-    ordpix = None
-    # Prepare and fit the sky background pixels in every order
-    msgs.work("Multiprocess this step to make it faster")
-    skybg = np.zeros_like(sciframe)
-    for o in range(norders):
-        xpix, ypix = np.where(ordpix==1+o)
-        msgs.info("Preparing sky pixels in order {0:d}/{1:d} for a b-spline fit".format(o+1,norders))
-        #xbarr, ybarr = cybspline.prepare_bsplfit(arc, pixmap, tilts, xmod, ycen, xpix, ypix)
-        msgs.info("Performing b-spline fir to oversampled sky background in order {0:d}/{1:d}".format(o+1,norders))
-        #ncoeff, k = flt.shape[0], 1
-        #mod_yarr = cybspline.bspline_fit(xmod, xbarr, ybarr, ebarr, min(np.min(xbarr),xstr), max(np.max(xbarr),xfin), ncoeff, k)
-        #skybg += cybspline.bspline_fitmod(xbarr, ybarr, ebarr, min(np.min(xbarr),xstr), max(np.max(xbarr),xfin), ncoeff, k, pixmap, tilts, xmod, ycen, xpix, ypix)
-
-    # Subtract the background
-    msgs.info("Subtracting the sky background from the science frame")
-    return sciframe-skybg, skybg
+    word = np.where((slf._slitpix[det - 1] == slitn + 1) & (slf._scimask[det - 1] == 0))
+    if word[0].size == 0:
+        msgs.warn("There are no pixels in slit {0:d}".format(slitn))
+        return np.zeros_like(sciframe)
+    xedges, modvals = object_profile(slf, sciframe, slitn, det, refine=refine, factor=3)
+    bincent = 0.5*(xedges[1:]+xedges[:-1])
+    npix = slf._pixwid[det - 1][slitn]
+    tilts = slf._tilts[det - 1].copy()
+    lordloc = slf._lordloc[det - 1][:, slitn]
+    rordloc = slf._rordloc[det - 1][:, slitn]
+    spatval = (word[1] - lordloc[word[0]] + refine) / (rordloc[word[0]] - lordloc[word[0]])
+    # Cumulative sum and normalize
+    csum = np.cumsum(modvals)
+    csum -= csum[0]
+    csum /= csum[-1]
+    # Find where the background begins
+    argl = np.argmin(np.abs(csum - 0.05))
+    argr = np.argmin(np.abs(csum - 0.95))
+    wl = np.where((modvals[1:] < modvals[:-1]) & (bincent[1:] < bincent[argl]))
+    wr = np.where((modvals[1:] > modvals[:-1]) & (bincent[1:] > bincent[argr]))
+    nl, nr = 0, 0
+    if wl[0].size != 0:
+        nl = np.max(wl[0])
+    if wr[0].size != 0:
+        nr = npix - np.min(wr[0])
+    if nl+nr < 5:
+        msgs.warn("The object profile appears to extrapolate to the edge of the detector")
+        msgs.info("A background subtraction will not be performed for slit".format(slitn+1))
+        return np.zeros_like(sciframe)
+    # Find background pixels and fit
+    wbgpix = np.where((spatval <= float(nl)/npix) | (spatval >= float(nr)/npix))
+    if settings.argflag['reduce']['skysub']['method'].lower() == 'bspline':
+        msgs.info("Using bspline sky subtraction")
+        srt = np.argsort(tilts[wbgpix])
+        ivar = (varframe > 0.)/(varframe + (varframe == 0))
```

Comment:
OK, good idea!  I've generated a new function in arutils to do just this.... but how about:

`ivar = (varframe > 0.) / (np.abs(varframe) + (varframe == 0))`